File: packages/java/src/main/java/io/constellationnetwork/metagraph/sdk/NetworkTypes.java
package io.constellationnetwork.metagraph.sdk;

/**
 * Network types for L1 client operations.
 */
public final class NetworkTypes {

    private NetworkTypes() {
        // Utility class
    }

    /**
     * Transaction status in the network.
     */
    public enum TransactionStatus {
        Waiting,
        InProgress,
        Accepted
    }

    /**
     * Pending transaction response from L1.
     */
    public static class PendingTransaction {
        private final String hash;
        private final TransactionStatus status;
        private final CurrencyTypes.CurrencyTransaction transaction;

        public PendingTransaction(String hash, TransactionStatus status, CurrencyTypes.CurrencyTransaction transaction) {
            this.hash = hash;
            this.status = status;
            this.transaction = transaction;
        }

        public String getHash() {
            return hash;
        }

        public TransactionStatus getStatus() {
            return status;
        }

        public CurrencyTypes.CurrencyTransaction getTransaction() {
            return transaction;
        }
    }

    /**
     * Response from posting a transaction.
     */
    public static class PostTransactionResponse {
        private final String hash;

        public PostTransactionResponse(String hash) {
            this.hash = hash;
        }

        public String getHash() {
            return hash;
        }
    }

    /**
     * Response from estimating data transaction fee.
     */
    public static class EstimateFeeResponse {
        private final long fee;
        private final String address;

        public EstimateFeeResponse(long fee, String address) {
            this.fee = fee;
            this.address = address;
        }

        public long getFee() {
            return fee;
        }

        public String getAddress() {
            return address;
        }
    }

    /**
     * Response from posting data.
     */
    public static class PostDataResponse {
        private final String hash;

        public PostDataResponse(String hash) {
            this.hash = hash;
        }

        public String getHash() {
            return hash;
        }
    }

    /**
     * Network error with status code and response details.
     */
    public static class NetworkException extends RuntimeException {
        private final int statusCode;
        private final String response;

        public NetworkException(String message) {
            this(message, 0, null);
        }

        public NetworkException(String message, int statusCode, String response) {
            super(message);
            this.statusCode = statusCode;
            this.response = response;
        }

        public NetworkException(String message, Throwable cause) {
            super(message, cause);
            this.statusCode = 0;
            this.response = null;
        }

        public int getStatusCode() {
            return statusCode;
        }

        public String getResponse() {
            return response;
        }

        @Override
        public String toString() {
            if (statusCode > 0) {
                return getMessage() + " (status: " + statusCode + ")";
            }
            return getMessage();
        }
    }
}

File: packages/go/network_types.go
package constellation

import (
	"errors"
	"fmt"
)

// NetworkConfig holds configuration for connecting to L1 nodes
type NetworkConfig struct {
	// L1URL is the Currency L1 endpoint URL (e.g., "http://localhost:9010")
	L1URL string
	// DataL1URL is the Data L1 endpoint URL (e.g., "http://localhost:8080")
	DataL1URL string
	// Timeout is the request timeout in seconds (default: 30)
	Timeout int
}

// RequestOptions holds options for individual requests
type RequestOptions struct {
	// Timeout is the request timeout in seconds
	Timeout int
}

// TransactionStatus represents the status of a transaction in the network
type TransactionStatus string

const (
	StatusWaiting    TransactionStatus = "Waiting"
	StatusInProgress TransactionStatus = "InProgress"
	StatusAccepted   TransactionStatus = "Accepted"
)

// PendingTransaction represents a pending transaction response from L1
type PendingTransaction struct {
	// Hash is the transaction hash
	Hash string `json:"hash"`
	// Status is the current status
	Status TransactionStatus `json:"status"`
	// Transaction is the transaction data
	Transaction CurrencyTransaction `json:"transaction"`
}

// PostTransactionResponse is the response from posting a transaction
type PostTransactionResponse struct {
	// Hash is the transaction hash
	Hash string `json:"hash"`
}

// EstimateFeeResponse is the response from estimating data transaction fee
type EstimateFeeResponse struct {
	// Fee is the estimated fee in smallest units
	Fee int64 `json:"fee"`
	// Address is the fee destination address
	Address string `json:"address"`
}

// PostDataResponse is the response from posting data
type PostDataResponse struct {
	// Hash is the data hash
	Hash string `json:"hash"`
}

// NetworkError represents a network operation error
type NetworkError struct {
	Message    string
	StatusCode int
	Response   string
}

func (e *NetworkError) Error() string {
	if e.StatusCode > 0 {
		return fmt.Sprintf("%s (status: %d)", e.Message, e.StatusCode)
	}
	return e.Message
}

// NewNetworkError creates a new NetworkError
func NewNetworkError(message string, statusCode int, response string) *NetworkError {
	return &NetworkError{
		Message:    message,
		StatusCode: statusCode,
		Response:   response,
	}
}

// Common network errors
var (
	ErrL1URLRequired     = errors.New("L1URL is required for CurrencyL1Client")
	ErrDataL1URLRequired = errors.New("DataL1URL is required for DataL1Client")
	ErrRequestTimeout    = errors.New("request timeout")
)
